src/lamp_py/ingestion/glides.py

```diff
@@ -0,0 +1,405 @@
+                ("id", pyarrow.string()),
+                ("type", pyarrow.string()),
+                ("time", pyarrow.timestamp("ms", tz="UTC")),
+                ("source", pyarrow.string()),
+                ("specversion", pyarrow.string()),
+                ("dataschema", pyarrow.string()),
+            ]
+        )
+
+    @property
+    def unique_key(self) -> str:
+        return "tripUpdates"
+
+    def convert_records(self) -> pd.Dataset:
+        def flatten_multitypes(record: Dict) -> Dict:
+            """
+            For each update in a record, flatten out the objects in "cars",
+            "dropped", and "cars". These objects are poorly structured for our
+            needs and can't be strongly typed. Leave them as json strings for
+            now and parse them later when analyzing.
+            """
+            for update in record["data"]["tripUpdates"]:
+                # convert these objects into strings if they exist
+                for key in ["scheduled", "dropped", "cars"]:
+                    try:
+                        update[key] = str(update[key])
+                    except KeyError:
+                        pass
+
+            return record
+
+        process_logger = ProcessLogger(
+            process_name="convert_records", type=self.type
+        )
+        process_logger.log_start()
+
+        modified_records = [flatten_multitypes(r) for r in self.records]
+        tu_table = pyarrow.Table.from_pylist(
+            modified_records, schema=self.event_schema
+        )
+        tu_table = flatten_schema(tu_table)
+        tu_table = explode_table_column(tu_table, "data.tripUpdates")
+        tu_table = flatten_schema(tu_table)
+        tu_dataset = pd.dataset(tu_table)
+
+        process_logger.log_complete()
+        return tu_dataset
+
+
+def ingest_glides_events(
+    kinesis_reader: KinesisReader, metadata_queue: Queue[Optional[str]]
+) -> None:
+    """
+    ingest glides records from the kinesis stream and add them to parquet files
+    """
+    process_logger = ProcessLogger(process_name="ingest_glides_events")
+    process_logger.log_start()
+
+    converters = [
+        EditorChanges(),
+        OperatorSignIns(),
+        TripUpdates(),
+    ]
+
+    for record in kinesis_reader.get_records():
+        try:
+            # format this so it can be used to partition parquet files
+            record["time"] = datetime.fromisoformat(
+                record["time"].replace("Z", "+00:00")
+            )
+
+            data_keys = record["data"].keys()
+
+            converter_found = False
+            for converter in converters:
+                if converter.unique_key in data_keys:
+                    converter.records.append(record)
+                    converter_found = True
+                    break
+
+            if not converter_found:
+                raise KeyError(f"No distinguishing key in {data_keys}")
+        except Exception as e:
+            process_logger.log_failure(e)
```

Comment:
this is a little different from our normal way of logging failures where the process logger will have a start, potentially many failures, and a complete. it avoids making a subprocess logger.

happy to revert if you prefer.